Use the same config file as Tox

ryanhiebert opened the pull request:
If successful, this fixes #57.

No new tests have been added, though some have been modified to adjust to the new call signature.

ryanhiebert commented:
@giginet: Are you up for giving this a try, to see if works for you?

rpkilby commented:
Hi @ryanhiebert - I have a few comments and was looking at this yesterday, but got pulled away. Just letting you know that I've got feedback incoming.

rpkilby commented on src/tox_travis/after.py:
```diff
@@ -81,10 +80,9 @@ def travis_after():
     """Determine if this job should wait for the others."""
-    config = py.iniconfig.IniConfig('tox.ini')
-    section = config.sections.get('travis:after', {})
+    section = cfg.sections.get('travis:after', {})
```
I don't particularly like passing the tox config all the way down the call stack, but it seems kind of necessary in the current design. Other options would be to:
- set a module-level global that `after_config_matches` accesses directly.
- create a `ToxAfter` class that takes the config in it's init, then refactor the various module functions as methods, and `after_config_matches` accesses the `config` directly on the instance.

But I'm not convinced that these options are a clear improvement. Just ideas.

ryanhiebert replied:
I agree that's its a bit annoying to pass the config down the stack, but I also agree that I'm not convinced that the other options are clear improvements at this point. If we had other global(ish) state that we needed, then `ToxAfter` (and probably a matching `ToxEnv`) would probably be my preferred approach. As it is, we'd probably just end up trading `config` in several places with `self` everywhere and extra indentation.

rpkilby replied:
> As it is, we'd probably just end up trading config in several places with self everywhere and extra indentation.

exactly.

rpkilby commented on src/tox_travis/hooks.py:
```diff
@@ -1,23 +1,28 @@
 
 
-@tox.hookimpl
+@tox.hookimpl(trylast=True)
```
Given that the PR parses the ini anyway, is this necessary?

ryanhiebert replied:
Yep, as you already discovered.

rpkilby commented on src/tox_travis/utils.py:
```diff
@@ -30,3 +34,53 @@ def parse_dict(value):
+    Figure out the configfile that Tox is going to use, and
+    create a ``py.iniconfig.IniConfig`` from that file, or
+    ``None`` if no configuration file could be found.
+    """
```
Your docs are always great... 🙌

ryanhiebert commented on src/tox_travis/utils.py:
```diff
@@ -30,3 +34,53 @@ def parse_dict(value):
+    return config._cfg
+
+
+def get_section(section, cfg):
```
Looks like it. I thought that I'd be using this all over, but when I went through, no other place seemed like it would benefit from the code churn, so I abandoned it.

ryanhiebert commented on src/tox_travis/hooks.py:
```diff
@@ -1,23 +1,28 @@
+        travis_after_monkeypatch(cfg)
 
-    override_ignore_outcome(config)
+    override_ignore_outcome(config, cfg)
```
I agree that it's confusing. The competing goal which motivated using `cfg` instead of `config` is that 1) almost nothing needs the full config, and 2) `config._cfg` is obviously an implementation detail of `config`, and I'd rather not scatter unstable APIs everywhere. Still it's confusing, and I can concede that there's not enough places we have to work with it that it's too big a deal.

rpkilby replied:
> `config._cfg` is obviously an implementation detail of `config`, and I'd rather not scatter unstable APIs everywhere.

True, what makes sense to me would be to 
- hide the internal API in a utility method
- only pass config, and call the utility where necessary

something like:
```py
# utils.py
def get_iniconfig(config):
    return config._cfg

# wherever the IniConfig is needed
tox_ini = get_iniconfig(config)
```

ryanhiebert replied:
I'd be happy with that.

ryanhiebert commented:
I was wondering if this would be superceded by #61, but clearly it's based on this one. I'm going to clean this up and merge it. I'm not going to do a release yet, since there seems to be more stuff still in progress.

rpkilby commented:
Tests?

ryanhiebert commented:
Yes, those too. I spoke like I wasn't going to, but I definitely do, and hopefully a green review too ;-)

jayvdb commented on src/tox_travis/hooks.py:
```diff
@@ -1,23 +1,28 @@
+        travis_after_monkeypatch(cfg)
 
-    override_ignore_outcome(config)
+    override_ignore_outcome(config, cfg)
```
For this one, why not get the boolean `unignore_outcomes` here, and pass that to `override_ignore_outcome` as an arg.

I'd be tempted to do the same with `travis_after`, passing a `Section` or `env_requirements` list object down, so that the ini parsing is done here, which will make ini config errors happen earlier in the tox sequence.

ryanhiebert replied:
Way late, but FWIW I'm definitely considering this approach for a possible refactor. It mostly depends just how involved I need to get into the config. For unignore_outcomes, it's pretty simple. For toxenv, it's unfortunately a bit more complicated bacause of a backward compatibility concern. But it seems like a good approach in general.

ryanhiebert commented:
This is ready for review. I've added a test, and rebased it to be based on us now using the tox_configure hook. So it's just using the config that Tox loaded, instead of loading our own.

ryanhiebert commented:
>>> I'm going to clean this up and merge it.
>>
>> Tests?
>
> Yes, those too. I spoke like I wasn't going to, but I definitely do, and hopefully a green review too ;-)

Confession: I totally wanted to skip writing tests. My excuse was that it was a bug fix and I wanted to get it fixed quickly. In hindsight the original pull request was certainly too big to skip tests, and I'm really thankful to @rpkilby for calling me out on it with a single word. I apologize for obfuscating my true intentions at the time.

At this point the patch is much smaller, and has added a regression test. In an effort to maintain my own momentum, I may merge some PRs that I perceive to be less challenging, like this one now is, though I intend to give @rpkilby a couple days before I do that with this particular pull request. Or anyone else that's willing. I sometimes take @rpkilby for granted because of how absolutely awesome he's been as a maintainer.
